File: experimental/teams/csharp-dotnetcore/Scenarios/Cards/Cards.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Threading.Tasks;
using Microsoft.Bot.Schema;
using Microsoft.Bot.Schema.Teams;
using Newtonsoft.Json;

namespace Cards
{
    public static class Cards
    {
        public static O365ConnectorCard CreateSampleO365ConnectorCard()
        {
            var actionCard1 = new O365ConnectorCardActionCard(
                O365ConnectorCardActionCard.Type,
                "Multiple Choice",
                "card-1",
                new List<O365ConnectorCardInputBase>
                {
                    new O365ConnectorCardMultichoiceInput(
                        O365ConnectorCardMultichoiceInput.Type,
                        "list-1",
                        true,
                        "Pick multiple options",
                        null,
                        new List<O365ConnectorCardMultichoiceInputChoice>
                        {
                            new O365ConnectorCardMultichoiceInputChoice("Choice 1", "1"),
                            new O365ConnectorCardMultichoiceInputChoice("Choice 2", "2"),
                            new O365ConnectorCardMultichoiceInputChoice("Choice 3", "3")
                        },
                        "expanded",
                        true),
                    new O365ConnectorCardMultichoiceInput(
                        O365ConnectorCardMultichoiceInput.Type,
                        "list-2",
                        true,
                        "Pick multiple options",
                        null,
                        new List<O365ConnectorCardMultichoiceInputChoice>
                        {
                            new O365ConnectorCardMultichoiceInputChoice("Choice 4", "4"),
                            new O365ConnectorCardMultichoiceInputChoice("Choice 5", "5"),
                            new O365ConnectorCardMultichoiceInputChoice("Choice 6", "6")
                        },
                        "compact",
                        true),
                    new O365ConnectorCardMultichoiceInput(
                        O365ConnectorCardMultichoiceInput.Type,
                        "list-3",
                        false,
                        "Pick an option",
                        null,
                        new List<O365ConnectorCardMultichoiceInputChoice>
                        {
                            new O365ConnectorCardMultichoiceInputChoice("Choice a", "a"),
                            new O365ConnectorCardMultichoiceInputChoice("Choice b", "b"),
                            new O365ConnectorCardMultichoiceInputChoice("Choice c", "c")
                        },
                        "expanded",
                        false),
                    new O365ConnectorCardMultichoiceInput(
                        O365ConnectorCardMultichoiceInput.Type,
                        "list-4",
                        false,
                        "Pick an option",
                        null,
                        new List<O365ConnectorCardMultichoiceInputChoice>
                        {
                            new O365ConnectorCardMultichoiceInputChoice("Choice x", "x"),
                            new O365ConnectorCardMultichoiceInputChoice("Choice y", "y"),
                            new O365ConnectorCardMultichoiceInputChoice("Choice z", "z")
                        },
                        "compact",
                        false)
    },
                new List<O365ConnectorCardActionBase>
                {
                    new O365ConnectorCardHttpPOST(
                        O365ConnectorCardHttpPOST.Type,
                        "Send",
                        "card-1-btn-1",
                        @"{""list1"":""{{list-1.value}}"", ""list2"":""{{list-2.value}}"", ""list3"":""{{list-3.value}}"", ""list4"":""{{list-4.value}}""}")
                });

            var actionCard2 = new O365ConnectorCardActionCard(
                O365ConnectorCardActionCard.Type,
                "Text Input",
                "card-2",
                new List<O365ConnectorCardInputBase>
                {
                    new O365ConnectorCardTextInput(
                        O365ConnectorCardTextInput.Type,
                        "text-1",
                        false,
                        "multiline, no maxLength",
                        null,
                        true,
                        null),
                    new O365ConnectorCardTextInput(
                        O365ConnectorCardTextInput.Type,
                        "text-2",
                        false,
                        "single line, no maxLength",
                        null,
                        false,
                        null),
                    new O365ConnectorCardTextInput(
                        O365ConnectorCardTextInput.Type,
                        "text-3",
                        true,
                        "multiline, max len = 10, isRequired",
                        null,
                        true,
                        10),
                    new O365ConnectorCardTextInput(
                        O365ConnectorCardTextInput.Type,
                        "text-4",
                        true,
                        "single line, max len = 10, isRequired",
                        null,
                        false,
                        10)
                },
                new List<O365ConnectorCardActionBase>
                {
                    new O365ConnectorCardHttpPOST(
                        O365ConnectorCardHttpPOST.Type,
                        "Send",
                        "card-2-btn-1",
                        @"{""text1"":""{{text-1.value}}"", ""text2"":""{{text-2.value}}"", ""text3"":""{{text-3.value}}"", ""text4"":""{{text-4.value}}""}")
                });

            var actionCard3 = new O365ConnectorCardActionCard(
                O365ConnectorCardActionCard.Type,
                "Date Input",
                "card-3",
                new List<O365ConnectorCardInputBase>
                {
                    new O365ConnectorCardDateInput(
                        O365ConnectorCardDateInput.Type,
                        "date-1",
                        true,
                        "date with time",
                        null,
                        true),
                    new O365ConnectorCardDateInput(
                        O365ConnectorCardDateInput.Type,
                        "date-2",
                        false,
                        "date only",
                        null,
                        false)
                },
                new List<O365ConnectorCardActionBase>
                {
                    new O365ConnectorCardHttpPOST(
                        O365ConnectorCardHttpPOST.Type,
                        "Send",
                        "card-3-btn-1",
                        @"{""date1"":""{{date-1.value}}"", ""date2"":""{{date-2.value}}""}")
                });

            var section = new O365ConnectorCardSection(
                "**section title**",
                "section text",
                "activity title",
                "activity subtitle",
                "activity text",
                "http://connectorsdemo.azurewebsites.net/images/MSC12_Oscar_002.jpg",
                "avatar",
                true,
                new List<O365ConnectorCardFact>
                {
                    new O365ConnectorCardFact("Fact name 1", "Fact value 1"),
                    new O365ConnectorCardFact("Fact name 2", "Fact value 2"),
                },
                new List<O365ConnectorCardImage>
                {
                    new O365ConnectorCardImage
                    {
                        Image = "http://connectorsdemo.azurewebsites.net/images/MicrosoftSurface_024_Cafe_OH-06315_VS_R1c.jpg",
                        Title = "image 1"
                    },
                    new O365ConnectorCardImage
                    {
                        Image = "http://connectorsdemo.azurewebsites.net/images/WIN12_Scene_01.jpg",
                        Title = "image 2"
                    },
                    new O365ConnectorCardImage
                    {
                        Image = "http://connectorsdemo.azurewebsites.net/images/WIN12_Anthony_02.jpg",
                        Title = "image 3"
                    }
                });

            O365ConnectorCard card = new O365ConnectorCard()
            {
                Summary = "O365 card summary",
                ThemeColor = "#E67A9E",
                Title = "card title",
                Text = "card text",
                Sections = new List<O365ConnectorCardSection> { section },
                PotentialAction = new List<O365ConnectorCardActionBase>
                    {
                        actionCard1,
                        actionCard2,
                        actionCard3,
                        new O365ConnectorCardViewAction(
                            O365ConnectorCardViewAction.Type,
                            "View Action",
                            null,
                            new List<string>
                            {
                                "http://microsoft.com"
                            }),
                        new O365ConnectorCardOpenUri(
                            O365ConnectorCardOpenUri.Type,
                            "Open Uri",
                            "open-uri",
                            new List<O365ConnectorCardOpenUriTarget>
                            {
                                new O365ConnectorCardOpenUriTarget
                                {
                                    Os = "default",
                                    Uri = "http://microsoft.com"
                                },
                                new O365ConnectorCardOpenUriTarget
                                {
                                    Os = "iOS",
                                    Uri = "http://microsoft.com"
                                },
                                new O365ConnectorCardOpenUriTarget
                                {
                                    Os = "android",
                                    Uri = "http://microsoft.com"
                                },
                                new O365ConnectorCardOpenUriTarget
                                {
                                    Os = "windows",
                                    Uri = "http://microsoft.com"
                                }
                            })
                    }
            };

            return card;
        }

        public static Attachment CreateAdaptiveCardAttachment()
        {
            // combine path for cross platform support
            string[] paths = { ".", "Resources", "adaptiveCard.json" };
            var adaptiveCardJson = File.ReadAllText(Path.Combine(paths));

            var adaptiveCardAttachment = new Attachment()
            {
                ContentType = "application/vnd.microsoft.card.adaptive",
                Content = JsonConvert.DeserializeObject(adaptiveCardJson),
            };
            return adaptiveCardAttachment;
        }

        public static HeroCard GetHeroCard()
        {
            var heroCard = new HeroCard
            {
                Title = "BotFramework Hero Card",
                Subtitle = "Microsoft Bot Framework",
                Text = "Build and connect intelligent bots to interact with your users naturally wherever they are," +
                       " from text/sms to Skype, Slack, Office 365 mail and other popular services.",
                Images = new List<CardImage> { new CardImage("https://sec.ch9.ms/ch9/7ff5/e07cfef0-aa3b-40bb-9baa-7c9ef8ff7ff5/buildreactionbotframework_960.jpg") },
                Buttons = new List<CardAction> { new CardAction(ActionTypes.OpenUrl, "Get Started", value: "https://docs.microsoft.com/bot-framework") },
            };

            return heroCard;
        }

        public static ThumbnailCard GetThumbnailCard()
        {
            var heroCard = new ThumbnailCard
            {
                Title = "BotFramework Thumbnail Card",
                Subtitle = "Microsoft Bot Framework",
                Text = "Build and connect intelligent bots to interact with your users naturally wherever they are," +
                       " from text/sms to Skype, Slack, Office 365 mail and other popular services.",
                Images = new List<CardImage> { new CardImage("https://sec.ch9.ms/ch9/7ff5/e07cfef0-aa3b-40bb-9baa-7c9ef8ff7ff5/buildreactionbotframework_960.jpg") },
                Buttons = new List<CardAction> { new CardAction(ActionTypes.OpenUrl, "Get Started", value: "https://docs.microsoft.com/bot-framework") },
            };

            return heroCard;
        }

        public static ReceiptCard GetReceiptCard()
        {
            var receiptCard = new ReceiptCard
            {
                Title = "John Doe",
                Facts = new List<Fact> { new Fact("Order Number", "1234"), new Fact("Payment Method", "VISA 5555-****") },
                Items = new List<ReceiptItem>
                {
                    new ReceiptItem(
                        "Data Transfer",
                        price: "$ 38.45",
                        quantity: "368",
                        image: new CardImage(url: "https://github.com/amido/azure-vector-icons/raw/master/renders/traffic-manager.png")),
                    new ReceiptItem(
                        "App Service",
                        price: "$ 45.00",
                        quantity: "720",
                        image: new CardImage(url: "https://github.com/amido/azure-vector-icons/raw/master/renders/cloud-service.png")),
                },
                Tax = "$ 7.50",
                Total = "$ 90.95",
                Buttons = new List<CardAction>
                {
                    new CardAction(
                        ActionTypes.OpenUrl,
                        "More information",
                        "https://account.windowsazure.com/content/6.10.1.38-.8225.160809-1618/aux-pre/images/offer-icon-freetrial.png",
                        value: "https://azure.microsoft.com/en-us/pricing/"),
                },
            };

            return receiptCard;
        }

        public static SigninCard GetSigninCard()
        {
            var signinCard = new SigninCard
            {
                Text = "BotFramework Sign-in Card",
                Buttons = new List<CardAction> { new CardAction(ActionTypes.Signin, "Sign-in", value: "https://login.microsoftonline.com/") },
            };

            return signinCard;
        }
    }
}

File: samples/java_springboot/06.using-cards/src/main/java/com/microsoft/bot/sample/usingcards/Cards.java
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

package com.microsoft.bot.sample.usingcards;

import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.node.ObjectNode;
import com.microsoft.bot.schema.ActionTypes;
import com.microsoft.bot.schema.AnimationCard;
import com.microsoft.bot.schema.Attachment;
import com.microsoft.bot.schema.AudioCard;
import com.microsoft.bot.schema.CardAction;
import com.microsoft.bot.schema.CardImage;
import com.microsoft.bot.schema.Fact;
import com.microsoft.bot.schema.HeroCard;
import com.microsoft.bot.schema.MediaUrl;
import com.microsoft.bot.schema.OAuthCard;
import com.microsoft.bot.schema.ReceiptCard;
import com.microsoft.bot.schema.ReceiptItem;
import com.microsoft.bot.schema.SigninCard;
import com.microsoft.bot.schema.ThumbnailCard;
import com.microsoft.bot.schema.ThumbnailUrl;
import com.microsoft.bot.schema.VideoCard;
import java.io.InputStream;
import java.nio.charset.StandardCharsets;
import java.util.concurrent.CompletionException;
import org.apache.commons.io.IOUtils;

public class Cards {
    public static Attachment createAdaptiveCardAttachment() {
        Attachment adaptiveCardAttachment = new Attachment();

        try (
            InputStream inputStream = adaptiveCardAttachment.getClass().getClassLoader()
            .getResourceAsStream("adaptiveCard.json")
        ) {
            String result = IOUtils.toString(inputStream, StandardCharsets.UTF_8);

            adaptiveCardAttachment.setContentType("application/vnd.microsoft.card.adaptive");
            adaptiveCardAttachment.setContent(new ObjectMapper().readValue(result, ObjectNode.class));

            return adaptiveCardAttachment;
        } catch (Throwable t) {
            throw new CompletionException(t);
        }
    }

    public static HeroCard getHeroCard() {
        HeroCard heroCard = new HeroCard();
        heroCard.setTitle("BotFramework Hero Card");
        heroCard.setSubtitle("Microsoft Bot Framework");
        heroCard.setText("Build and connect intelligent bots to interact with your users naturally wherever they are," +
                    " from text/sms to Skype, Slack, Office 365 mail and other popular services.");
        heroCard.setImages(new CardImage("https://sec.ch9.ms/ch9/7ff5/e07cfef0-aa3b-40bb-9baa-7c9ef8ff7ff5/buildreactionbotframework_960.jpg"));
        heroCard.setButtons(new CardAction(ActionTypes.OPEN_URL, "Get Started", "https://docs.microsoft.com/bot-framework"));

        return heroCard;
    }

    public static ThumbnailCard getThumbnailCard() {
        ThumbnailCard thumbnailCard = new ThumbnailCard();
        thumbnailCard.setTitle("BotFramework Thumbnail Card");
        thumbnailCard.setSubtitle("Microsoft Bot Framework");
        thumbnailCard.setText("Build and connect intelligent bots to interact with your users naturally wherever they are," +
                    " from text/sms to Skype, Slack, Office 365 mail and other popular services.");
        thumbnailCard.setImages(new CardImage("https://sec.ch9.ms/ch9/7ff5/e07cfef0-aa3b-40bb-9baa-7c9ef8ff7ff5/buildreactionbotframework_960.jpg"));
        thumbnailCard.setButtons(new CardAction(
            ActionTypes.OPEN_URL, "Get Started", "https://docs.microsoft.com/bot-framework"));

        return thumbnailCard;
    }

    public static ReceiptCard getReceiptCard() {
        ReceiptCard receiptCard = new ReceiptCard();
        receiptCard.setTitle("John Doe");
        ReceiptItem receiptDataTransfer = new ReceiptItem();
        receiptDataTransfer.setTitle("Data Transfer");
        receiptDataTransfer.setPrice("$ 38.45");
        receiptDataTransfer.setQuantity("368");
        receiptDataTransfer.setImage(new CardImage("https://github.com/amido/azure-vector-icons/raw/master/renders/traffic-manager.png"));
        ReceiptItem receiptAppService = new ReceiptItem();
        receiptAppService.setTitle("App Service");
        receiptAppService.setPrice("$ 45.00");
        receiptAppService.setQuantity("720");
        receiptAppService.setImage(new CardImage("https://github.com/amido/azure-vector-icons/raw/master/renders/cloud-service.png"));
        receiptCard.setFacts(new Fact("Order Number", "1234"), new Fact("Payment Method", "VISA 5555-****"));
        receiptCard.setItems(receiptDataTransfer, receiptAppService);
        receiptCard.setTax("$ 7.50");
        receiptCard.setTotal("$ 90.95");
        CardAction cardAction = new CardAction(ActionTypes.OPEN_URL, "More information");
        cardAction.setImage("https://account.windowsazure.com/content/6.10.1.38-.8225.160809-1618/aux-pre/images/offer-icon-freetrial.png");
        cardAction.setValue("https://azure.microsoft.com/en-us/pricing/");
        receiptCard.setButtons(cardAction);

        return receiptCard;
    }

    public static SigninCard getSigninCard() {
        SigninCard signinCard = new SigninCard();
        signinCard.setText("BotFramework Sign-in Card");
        signinCard.setButtons(new CardAction(ActionTypes.SIGNIN, "Sign-in", "https://login.microsoftonline.com/"));
        return signinCard;
    }

    public static AnimationCard getAnimationCard() {
        AnimationCard animationCard = new AnimationCard();
        animationCard.setTitle("Microsoft Bot Framework");
        animationCard.setSubtitle("Animation Card");
        animationCard.setImage(new ThumbnailUrl("https://docs.microsoft.com/en-us/bot-framework/media/how-it-works/architecture-resize.png"));
        animationCard.setMedia(new MediaUrl("http://i.giphy.com/Ki55RUbOV5njy.gif"));
        return animationCard;
    }

    public static VideoCard getVideoCard() {
        VideoCard videoCard = new VideoCard();
        videoCard.setTitle("Big Buck Bunny");
        videoCard.setSubtitle("by the Blender Institute");
        videoCard.setText("Big Buck Bunny (code-named Peach) is a short computer-animated comedy film by the Blender Institute," +
            " part of the Blender Foundation. Like the foundation's previous film Elephants Dream," +
            " the film was made using Blender, a free software application for animation made by the same foundation." +
            " It was released as an open-source film under Creative Commons License Attribution 3.0.");
        videoCard.setImage(new ThumbnailUrl("https://upload.wikimedia.org/wikipedia/commons/thumb/c/c5/Big_buck_bunny_poster_big.jpg/220px-Big_buck_bunny_poster_big.jpg"));
        videoCard.setMedia(new MediaUrl("http://download.blender.org/peach/bigbuckbunny_movies/BigBuckBunny_320x180.mp4"));
        videoCard.setButtons(new CardAction(ActionTypes.OPEN_URL, "Learn More", "https://peach.blender.org/"));

        return videoCard;
    }

    public static AudioCard getAudioCard() {
        AudioCard audioCard = new AudioCard();
        audioCard.setTitle("I am your father");
        audioCard.setSubtitle("Star Wars: Episode V - The Empire Strikes Back");
        audioCard.setText("The Empire Strikes Back (also known as Star Wars: Episode V – The Empire Strikes Back)" +
                " is a 1980 American epic space opera film directed by Irvin Kershner. Leigh Brackett and" +
                " Lawrence Kasdan wrote the screenplay, with George Lucas writing the film's story and serving" +
                " as executive producer. The second installment in the original Star Wars trilogy, it was produced" +
                " by Gary Kurtz for Lucasfilm Ltd. and stars Mark Hamill, Harrison Ford, Carrie Fisher, Billy Dee Williams," +
                " Anthony Daniels, David Prowse, Kenny Baker, Peter Mayhew and Frank Oz.");
        audioCard.setImage(new ThumbnailUrl("https://upload.wikimedia.org/wikipedia/en/3/3c/SW_-_Empire_Strikes_Back.jpg"));
        audioCard.setMedia(new MediaUrl("http://www.wavlist.com/movies/004/father.wav"));
        audioCard.setButtons(new CardAction(ActionTypes.OPEN_URL, "Read More", "https://en.wikipedia.org/wiki/The_Empire_Strikes_Back"));

        return audioCard;
    }

    public static OAuthCard getOAuthCard() {
        OAuthCard oauthCard = new OAuthCard();
        oauthCard.setText("BotFramework OAuth Card");
        oauthCard.setConnectionName("OAuth connection"); // Replace with the name of your Azure AD connection.
        oauthCard.setButtons(new CardAction(ActionTypes.SIGNIN, "Sign In", "https://example.org/signin"));
        return oauthCard;
    }
}
